src/controllers/opportunities.js

```diff
@@ -0,0 +1,248 @@
+   * @param {Object} context of the request
+   * @return {Promise<Response>} Opportunity response.
+   */
+  const createOpportunity = async (context) => {
+    const siteId = context.params?.siteId;
+    if (!hasText(siteId)) {
+      return badRequest('Site ID required');
+    }
+    if (!isNonEmptyObject(context.data)) {
+      return badRequest('No data provided');
+    }
+
+    context.data.siteId = siteId;
+    try {
+      const oppty = await Opportunity.create(context.data);
+      return createResponse(OpportunityDto.toJSON(oppty), 201);
+    } catch (e) {
+      return badRequest(e.message);
+    }
+  };
+
+  /**
+   * Updates data for an opportunity
+   * @param {Object} context of the request
+   * @returns {Promise<Response>} the updated opportunity data
+   */
+  const patchOpportunity = async (context) => {
+    const siteId = context.params?.siteId;
+    const opportunityId = context.params?.opportunityId;
+    // validate parameters
+    if (!hasText(siteId)) {
+      return badRequest('Site ID required');
+    }
+    if (!hasText(opportunityId)) {
+      return badRequest('Opportunity ID required');
+    }
+
+    const opportunity = await Opportunity.findById(opportunityId);
+    if (!opportunity || opportunity.getSiteId() !== siteId) {
+      return notFound('Opportunity not found');
+    }
+    // validate request body
+    if (!isNonEmptyObject(context.data)) {
+      return badRequest('No updates provided');
+    }
+
+    // eslint-disable-next-line object-curly-newline
+    const { auditId, runbook, data, title, description, status, guidance, tags } = context.data;
+    // update opportunity with new data
+    let hasUpdates = false;
+    try {
+      if (auditId && auditId !== opportunity.getAuditId()) {
+        hasUpdates = true;
+        opportunity.setAuditId(auditId);
+      }
+      if (runbook && runbook !== opportunity.getRunbook()) {
+        hasUpdates = true;
+        opportunity.setRunbook(runbook);
+      }
+      if (isNonEmptyObject(data)) {
+        hasUpdates = true;
+        opportunity.setData(data);
+      }
+
+      if (title && title !== opportunity.getTitle()) {
+        hasUpdates = true;
+        opportunity.setTitle(title);
+      }
+      if (description && description !== opportunity.getDescription()) {
+        hasUpdates = true;
+        opportunity.setDescription(description);
+      }
+      if (status && status !== opportunity.getStatus()) {
+        hasUpdates = true;
+        opportunity.setStatus(status);
+      }
+      if (isNonEmptyObject(guidance)) {
+        hasUpdates = true;
+        opportunity.setGuidance(guidance);
+      }
+      if (tags && !arrayEquals(tags, opportunity.getTags())) {
+        hasUpdates = true;
+        opportunity.setTags(tags);
+      }
+      if (hasUpdates) {
+        const updatedOppty = await opportunity.save(opportunity);
+        return ok(OpportunityDto.toJSON(updatedOppty));
+      }
+    } catch (e) {
+      return badRequest(e.message);
```

Comment:
bad request would only be applicable if a validation error occurs, but there might also be other errors

Reply:
DONE
We need to decide what to do with data access errors. Do we return them as they come, or do we transform the message somehow?